spec/ics-027-interchain-account/README.md

```diff
@@ -0,0 +1,285 @@
+  // version not used at present  
+  abortTransactionUnless(version === "")  
+}  
+```
+
+```typescript  
+function onChanOpenTry(  
+  order: ChannelOrder,  
+  connectionHops: [Identifier],  
+  portIdentifier: Identifier,  
+  channelIdentifier: Identifier,  
+  counterpartyPortIdentifier: Identifier,  
+  counterpartyChannelIdentifier: Identifier,  
+  version: string,  
+  counterpartyVersion: string) {  
+  // only ordered channels allowed  
+  abortTransactionUnless(order === ORDERED)  
+  // version not used at present  
+  abortTransactionUnless(version === "")  
+  abortTransactionUnless(counterpartyVersion === "")  
+  // only allow channels to "interchain-account" port on counterparty chain  
+  abortTransactionUnless(counterpartyPortIdentifier === "interchain-account")  
+}  
+```
+
+```typescript  
+function onChanOpenAck(  
+  portIdentifier: Identifier,  
+  channelIdentifier: Identifier,  
+  version: string) {  
+  // version not used at present  
+  abortTransactionUnless(version === "")  
+  // port has already been validated  
+}  
+```
+
+```typescript  
+function onChanOpenConfirm(  
+  portIdentifier: Identifier,  
+  channelIdentifier: Identifier) {  
+  // accept channel confirmations, port has already been validated  
+}  
+```
+
+```typescript  
+function onChanCloseInit(  
+  portIdentifier: Identifier,  
+  channelIdentifier: Identifier) {  
+  // no action necessary  
+}  
+```
+
+```typescript  
+function onChanCloseConfirm(  
+  portIdentifier: Identifier,  
+  channelIdentifier: Identifier) {  
+  // no action necessary  
+}  
+```
+
+### **Packet relay**  
+
+In plain English, between chains `A` and `B`. It will describe only the case that chain A wants to register an Interchain account on chain B and control it. Moreover, this system can also be applied the other way around.  
+
+```typescript  
+function onRecvPacket(packet: Packet): bytes {  
+  if (packet.data is RunTxPacketData) {  
+    const tx = deserializeTx(packet.data.txBytes)
+    abortTransactionUnless(authenticateTx(tx) == true)
+    return runTx(tx)
+  }  
+      
+  if (packet.data is RegisterIBCAccountPacketData) {  
+    RegisterIBCAccountPacketData data = packet.data  
+    identifier = "{packet/sourcePort}/{packet.sourceChannel}"  
+    const address = generateAddress(identifier, packet.salt)
+    createAccount(address)
+    // Return generated address.
+    return address
+  }   
+      
+  return 0x  
+}  
+```
+
+```typescript  
+function onAcknowledgePacket(  
+  packet: Packet,  
+  acknowledgement: bytes) {  
+  if (packet.data is RegisterIBCAccountPacketData) {
+    if (acknowledgement !== 0x) {
+      identifier = "{packet/sourcePort}/{packet.sourceChannel}"
+      onAccountCreated(identifier, acknowledgement)
+    }
+  }
+  if (packet.data is RunTxPacketData) {
+    identifier = "{packet/destPort}/{packet.destChannel}"  
+    onTxSucceeded(identifier: Identifier, packet.data.tyBytes)
```

Comment:
```suggestion
    onTxSucceeded(identifier: Identifier, packet.data.txBytes)
```

Reply:
Shouldn't you pass the acknowledgement? It might be an error.

Reply:
The reason why passes tx bytes is because it let a chain sending interchain transaction to know which transaction is succeeded or failed. But, I think it is needed to pass the acknowledgement if interchain transaction is failed because sending chain should know why transaction is failed. It was my mistake.